File: main/largest-rectangle/largest-rectangle.cs
using System;
using System.Collections.Generic;
using System.Linq;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal static class Solution {
    private static readonly IEnumerable<long> Zero = Enumerable.Repeat(0L, 1);

    private static long ComputeMaxArea(this IEnumerable<long> heights)
    {
        var h = Zero.Concat(heights).Concat(Zero).ToArray();
        var lefts = new Stack<int>(h.Length);
        var acc = 0L;

        foreach (var right in Enumerable.Range(0, h.Length)) {
            while (lefts.Count != 0 && h[lefts.Peek()] > h[right]) {
                var left = lefts.Pop();

                var width = right;
                if (lefts.Count != 0) width -= 1 + lefts.Peek();

                acc = Math.Max(acc, h[left] * width);
            }

            lefts.Push(right);
        }

        return acc;
    }

    private static void Main()
    {
        Console.ReadLine(); // don't need n

        Console.WriteLine(Console.ReadLine().Split().Select(long.Parse)
                                                    .ComputeMaxArea());
    }
}

File: main/largest-rectangle/largest-rectangle.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Scanner;
import java.util.stream.IntStream;
import java.util.stream.LongStream;

final class Solution {
    private static LongStream zero()
    {
        return LongStream.of(0L);
    }
    
    private static LongStream zeroBracket(final LongStream values)
    {
        return LongStream.concat(LongStream.concat(zero(), values), zero());
    }
    
    private static long computeMaxArea(final LongStream heights)
    {
        final long[] h = zeroBracket(heights).toArray();
        final Deque<Integer> lefts = new ArrayDeque<>(h.length);
        long acc = 0L;
        
        for (int right = 0; right != h.length; ++right) {
            while (!lefts.isEmpty() && h[lefts.peek()] > h[right]) {
                final int left = lefts.pop();
                
                int width = right;
                if (!lefts.isEmpty()) width -= 1 + lefts.peek();
                
                acc = Math.max(acc, h[left] * width);
            }
            
            lefts.push(right);
        }
        
        return acc;
    }
    
    private static LongStream readRecord(final Scanner sc)
    {
        return IntStream.range(0, sc.nextInt()).mapToLong(i -> sc.nextLong());
    }
    
    public static void main(final String[] args) {
        try (final Scanner sc = new Scanner(System.in)) {
            System.out.println(computeMaxArea(readRecord(sc)));
        }
    }
    
    private Solution() { } // the Solution class should not be instantiated
}
